bson-codec-design/DESIGN.md

```diff
@@ -0,0 +1,157 @@
+restrictive because users sometimes want to handle the processing for types they do not define
+themselves. For example, users who generate types using a protocol buffer library
+or users who want to handle the processing of types in the standard library.
+
+The design of the v1 encoder and decoder for the Go driver presents problems when users wish to
+have more control over how specific types are handled when encoding or decoding. The control that
+users require exists on two main levels:
+- control over how individual types are encoded or decoded
+- control over how tags are processed
+
+The granularity of control users require can be defined in three levels:
+- Control over how individual types are encoded and decoded when they are embedded in another type,
+such as a map or a struct
+- Control over how individual types are encoded and decoded when they are provided directly to the
+encoder or decoder
+- Control over how tags are processed and the parameters that are returned
+
+The proposed design allows for the first two granularities to be handled together.
+
+## High Level Design
+### Registries
+To enable users to control how types are encoded and decoded, the proposed design uses a type
+registry. This registry will be provided in any context where encoding or decoding occurs. The registry
+will be used to lookup a type's codec, which will then be used for the actual encoding a decoding.
+This will allow users to control nearly all aspects of the encoding and decoding process. One
+failing of the v1 design is that users cannot control the encoding and decoding of types
+they do not own, e.g. `time.Time`. The proposed design handles this by allowing users to register a
+codec for any type.
+
+Pointer and value types are handled separately in the registry as this simplifies the registry and
+the implementations of codecs, since they will only have to worry about a single type and not the
+type plus a pointer to that type. While likely not useful, there is no design
+reason to prevent users from registering codecs for internal types like `bson.Element`,
+`bson.Document`, or `bson.Reader`. Users can override the functionality of built in types as
+well. The registry will be used for all encoding and decoding operations, since the cost of calling
+`reflect.TypeOf` is inexpensive.
+
+#### No Global Registries
+The proposed design does not contain a global type nor a global interface registry. This is mainly
+because [package level variables are generally discouraged in
+Go](http://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html). Instead, registries are
+explicitly passed to the things that require them. For instance, when constructing an Encoder or a
+Decoder the registry is a parameter. The Marshal and Unmarshal global package functions will be
+altered and instead there will be Marshal and Unmarshal methods on the Registry type. The new
+Marshal and Unmarshal package functions will take a Registry as a parameter.
```

Comment:
I think that its desirable to have a default registry, so that `bson.Marshal` and `bson.Unmarshal` have the same signature as `mgo/bson` *and* `encoding/json`.

Reply:
`mgo/bson` and `encoding/json` don't allow their `Marshal` and `Unmarshal` functions to be altered by a global registry, so I don't think the "It should look the same" argument applies here. The aforementioned packages operate on a static set of variables, while this design lays out a dynamic set of variables.

Reply:
I think it's fine to provide additional functionality to support additional use-cases (with different/additional APIs), but I think the base functionality needs to look and more importantly be able to operate the same.